<a href="https://colab.research.google.com/github/Fabian830348/Bases_Datos/blob/master/Notebook_2_DataFrames_Parte1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<table>
    <tr>
        <td><img src="https://s3.amazonaws.com/media-p.slid.es/uploads/1485763/images/9060062/Header.png" width="300"/></td>
        <td>&nbsp;</td>
        <td>
            <h1 style="font-size:200%;color:blue;text-align:center">    <FONT COLOR="blue">  Python </p> DataFrames </p> Parte 1   </FONT>         </h1></td>         
        <td>
            <tp><p style="font-size:99%;text-align:center">Python Básico </p></tp>
            <tp><p style="font-size:115%;text-align:center">Maestría MACC 2025-1</p></tp>
            <tp><p style="font-size:115%;text-align:center">Prof. Fabián Sánchez</p></tp>
        </td>
    </tr>
</table>

# <FONT SIZE=5 COLOR="green"> 1. Definición de DataFrame en Python </FONT>

- En este notebook trabajaremos uno de los objetos más importante cuando manipulamos y procesamos datos, los DataFrames.

- Recordemos que existen dos grupos grandes de datos que se denominan:

  - **Datos estructurados**: son aquellos que se pueden representar en un arreglo bidimensional: *fila-columna*. Es decir, las clásicas hojas de cálculo de excel.
  - **Datos no estructurados**: son los que no podemos representar en forma estructurada *fila-columna*. Por ejemplo, archivos de vídeo, audio, texto, imágenes, etc.


 De acuerdo con lo anterior definimos:

 ***Dataframe***: son arreglos o estructuras bidimensionales compuestos por fila y por columnas, y cuyos datos pueden ser de diferentes tipos: números, cadenas de texto, booleanos, etc.

   - Un dataframe tiene una forma matricial donde se identifican filas y columnas. Soporta diferentes tipos de datos.
   - Un dataframe está compuesto por vectores (columnas) cada una de las cuales describe una atributo o característica (variable) de una población.
   - La librería más usada para manipulación de dataframes se denomina ***pandas*** en Python. Si trabajas en *RStudio* es ***dplyr***

<br>
<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/DATAFRAME.png?raw=true" alt="centered image" width="600" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>




Iniciaremos importando la librería *pandas*

In [ ]:
import pandas as pd
import numpy as np

# <FONT SIZE=5 COLOR="green"> 2. Generación de un DataFrame </FONT>

En esta sección revisaremos cómo se genera un dataframe a partir de:

- Listas
- Un diccionario

para generar un dataframe usamos la instrucción de pandas

$$pd.DataFrame()$$

## <FONT SIZE=4 COLOR="magenta"> 2.1 A partir de una Lista </FONT>

In [ ]:
lista1 = ["Fabian" , 30 , 164 , "Doctorado"]
lista2 = ["Camilo" , 24 , 174 , "Maestría"]
lista3 = ["Ana María" , 20 , 180 , "Profesional"]
lista4 = ["Andrea" , 36 , 154 , "Doctorado"]
lista5 = ["Mariana" , 45 , 164 , "Doctorado"]

pd.DataFrame([lista1, lista2, lista3, lista4, lista5])

observe que enumera las columnas pero no tiene nombres.

Ahora, vamos a colocar nombre a las columnas

In [ ]:
lista1 = ["Fabian" , 30 , 164 , "Doctorado"]
lista2 = ["Camilo" , 24 , 174 , "Maestría"]
lista3 = ["Ana María" , 20 , 180 , "Profesional"]
lista4 = ["Andrea" , 36 , 154 , "Doctorado"]
lista5 = ["Mariana" , 45 , 164 , "Doctorado"]

df1 = pd.DataFrame([lista1, lista2, lista3, lista4, lista5],                      # datos
              columns = ["Nombre", "Edad", "Estatura(cm)", "Nivel_Educativo"])        # variables

In [ ]:
df1

Ahora bien, si se tienen la listas que representan las columnas, el dataframe se construye de la siguiente forma

In [ ]:
lista1 = [30,40,50,60]
lista2 = ["Audi", "BMW", "Renault", "ByD"]
lista3 = [1.76, 1.50, 1.60, 1.80]
lista4 = ["Carlos", "Juan", "Hugo", "Ana"]

df2 =pd.DataFrame({"Edad" : lista1,
              "Marca_Auto" : lista2,
              "Estatura(mts)" : lista3,
              "Nombre" : lista4})
df2


## <FONT SIZE=4 COLOR="magenta"> 2.2 A partir de un diccionario </FONT>

Para construir el DataFrame a partir de un diccionario se usa

$$pd.DataFrame(diccionario)$$

In [ ]:
diccionario = {"Nombre" : ["Hugo", "Andres", "Camilo", "Ana" ],
               "Edad" : [70, 40, 24, 10],
               "Estatura" : [180, 170, 150, 176],
               "Nivel_Educativo" : ["Doctorado", "Profesional", "Maestría","Maestría"]}

df3 = pd.DataFrame(diccionario)
df3

In [ ]:
nombres = ["Hugo", "Andres", "Camilo", "Ana"]
edad =  [70, 40, 24, 10]
pd.DataFrame({"nombres" : nombres , "edad" : edad})


# <FONT SIZE=5 COLOR="green"> 3. Instrucciones básicas</FONT>

Algunas instrucciones básicas de un dataframe:

**df.info()** : Devuelve información (número de filas, número de columnas, índices, tipo de las columnas y memoria usado) sobre el DataFrame df.

**df.shape** : Devuelve una pareja con el número de filas y columnas del DataFrame df.

**df.size** : Devuelve el número de elementos del DataFrame.

**df.columns** : Devuelve una lista con los nombres de las columnas del DataFrame df.

**df.index** : Devuelve una lista con los nombres de las filas del DataFrame df.

**df.dtypes** : Devuelve una serie con los tipos de datos de las columnas del DataFrame df.

**df.head(n)** : Devuelve las n primeras filas del DataFrame df.

**df.tail(n)** : Devuelve las n últimas filas del DataFrame df.

**df.append({})** : para agregar una nueva fila al dataframe

# <FONT SIZE=5 COLOR="Green"> 4. Carga de Datos </FONT></center>

Vamos a cargar datos (dataframe) de dos formas diferentes:

  - Desde ***googledrive()***
  - Desde un repositorio externo. Generalmente GitHub.

## <FONT SIZE=4 COLOR="magenta"> 4.1 Usando Google-Drive </FONT></center>

- Para conectarse al drive usamos la siguiente instrucción

In [ ]:
# instrucción para conectarse con drive
from google.colab import drive
drive.mount('/content/drive')

Siempre que trabajemos con dataframes es importante cargar la librería pandas.

In [ ]:
#para importar la librería pandas.
import pandas as pd

Para cargar los archivos buscamos la ubicación de los mismos y la ruta en la parte izquierda del notebook

In [ ]:
ruta_drive = "/content/drive/MyDrive/Intro_python_2025/APARTAMENTOS.csv"

Como el archivo está en formato *.csv*, que es un archivo separado por coma usamos la función

$$pd.read\_csv(link)$$

In [ ]:
aptos = pd.read_csv(ruta_drive)
aptos

Unnamed: 0  precio     mt2 ubicacion  estrato  alcobas  banos balcon  \
0             1    79.0   43.16     norte        3        3      1     si   
1             2    93.0   56.92     norte        2        2      1     si   
2             3   100.0   66.40     norte        3        2      2     no   
3             4   123.0   61.85     norte        2        3      2     si   
4             5   135.0   89.80     norte        4        3      2     si   
..          ...     ...     ...       ...      ...      ...    ...    ...   
689         690   498.0  180.00  laureles        5        4      3     si   
690         691   508.0  209.00  laureles        6        4      3     si   
691         692   550.0  149.00  laureles        6        3      4     no   
692         693   630.0  179.00  laureles        5        3      3     si   
693         694  1350.0  494.00  laureles        5        3      4     si   

    parqueadero  administracion      avaluo terminado  
0            si           0.050   14.923002        no  
1            si           0.069   27.000000        si  
2            no           0.000   15.738427        no  
3            si           0.130   27.000000        no  
4            no           0.000   39.567000        si  
..          ...             ...         ...       ...  
689          si           0.570  233.784000        si  
690          si           0.450  258.850000        si  
691          si           0.747  326.157000        si  
692          si           0.340  324.114000        si  
693          si           1.600  320.000000        si  

[694 rows x 12 columns]

Si el archivo está en formato *.xlsx*, usamos la función

$$pd.read\_excel(link)$$

In [ ]:
# traer el conjunto de datos hepatitis
hepatitis = pd.read_excel("/content/drive/MyDrive/Intro_python_2025/HEPATITIS.xlsx")
hepatitis

Clase  Edad    Sexo Esteroides Antivirales Fatiga Malestar Anorexia  \
0      Vivo    30   Mujer         No          Sí     Sí       Sí       Sí   
1      Vivo    50  Hombre         No          Sí     No       Sí       Sí   
2      Vivo    78  Hombre         Sí          Sí     No       Sí       Sí   
3      Vivo    31  Hombre                     No     Sí       Sí       Sí   
4      Vivo    34  Hombre         Sí          Sí     Sí       Sí       Sí   
..      ...   ...     ...        ...         ...    ...      ...      ...   
147  Muerto    46  Hombre         Sí          Sí     No       No       No   
148    Vivo    44  Hombre         Sí          Sí     No       Sí       Sí   
149    Vivo    61  Hombre         No          Sí     No       No       Sí   
150    Vivo    53   Mujer         No          Sí     No       Sí       Sí   
151  Muerto    43  Hombre         Sí          Sí     No       Sí       Sí   

    Higado Grande Cirrosis Bazo palpable Arañas Ascites Varices  Billirubina  \
0              No       Sí            Sí     Sí      Sí      Sí          1.0   
1              No       Sí            Sí     Sí      Sí      Sí          0.9   
2              Sí       Sí            Sí     Sí      Sí      Sí          0.7   
3              Sí       Sí            Sí     Sí      Sí      Sí          0.7   
4              Sí       Sí            Sí     Sí      Sí      Sí          1.0   
..            ...      ...           ...    ...     ...     ...          ...   
147            Sí       Sí            Sí     No      No      No          7.6   
148            Sí       No            Sí     Sí      Sí      Sí          0.9   
149            No       No            Sí     No      Sí      Sí          0.8   
150            Sí       Sí            No     No      Sí      No          1.5   
151            Sí       Sí            No     No      No      Sí          1.2   

     Fosfato  SGOT  Albumina  Protina Histologia  
0         85    18       4.0        0         No  
1        135    42       3.5        0         No  
2         96    32       4.0        0         No  
3         46    52       4.0       80         No  
4          0   200       4.0        0         No  
..       ...   ...       ...      ...        ...  
147        0   242       3.3       50         Sí  
148      126   142       4.3        0         Sí  
149       75    20       4.1        0         Sí  
150       81    19       4.1       48         Sí  
151      100    19       3.1       42         Sí  

[152 rows x 20 columns]

Si el archivo está en formato *.json*, usamos la función

$$pd.read\_json(link)$$

In [ ]:
# traer el conjunto de datos .json
espectro = pd.read_json("/content/drive/MyDrive/Intro_python_2025/espectro.json")
espectro

nombre_sonda     ciudad_departamento  \
0                  Bogota_BOSA_2           Bogota_BOGOTA   
1           Santa_Marta_TERMINAL   Santa_Marta_MAGDALENA   
2    Zipaquira_ZIPA_CONSTRUCTORA  Zipaquira_CUNDINAMARCA   
3                 Armenia_CENTRO         Armenia_QUINDIO   
4             Pereira_CENTENARIO       Pereira_RISARALDA   
..                           ...                     ...   
995                Soacha_CENTRO     Soacha_CUNDINAMARCA   
996         Santa_Marta_TERMINAL   Santa_Marta_MAGDALENA   
997      Barranquilla_SAN_FELIPE  Barranquilla_ATLANTICO   
998               Pereira_BERLIN       Pereira_RISARALDA   
999           Bogota_LISBOA_SUBA           Bogota_BOGOTA   

              fecha_medicion  nivel_exposicion_campos  limite_restrictivo  \
0    2020-06-30T23:54:00.000                     1.11                  28   
1    2020-06-30T23:54:00.000                     2.14                  28   
2    2020-06-30T23:54:00.000                     1.53                  28   
3    2020-06-30T23:54:00.000                     1.51                  28   
4    2020-06-30T23:54:00.000                     0.00                  28   
..                       ...                      ...                 ...   
995  2020-06-30T22:06:00.000                     1.93                  28   
996  2020-06-30T22:06:00.000                     2.06                  28   
997  2020-06-30T22:06:00.000                     2.11                  28   
998  2020-06-30T22:06:00.000                     1.80                  28   
999  2020-06-30T22:06:00.000                     0.49                  28   

     porcentaje_exposicion                                    geocoded_column  
0                     3.96  {'type': 'Point', 'coordinates': [-74.1688982,...  
1                     7.64  {'type': 'Point', 'coordinates': [-74.18208333...  
2                     5.46  {'type': 'Point', 'coordinates': [-74.00063, 5...  
3                     5.39  {'type': 'Point', 'coordinates': [-75.67364, 4...  
4                     0.00  {'type': 'Point', 'coordinates': [-75.698193, ...  
..                     ...                                                ...  
995                   6.89  {'type': 'Point', 'coordinates': [-74.220488, ...  
996                   7.36  {'type': 'Point', 'coordinates': [-74.18208333...  
997                   7.54  {'type': 'Point', 'coordinates': [-74.80913888...  
998                   6.43  {'type': 'Point', 'coordinates': [-75.68085, 4...  
999                   1.75  {'type': 'Point', 'coordinates': [-74.117347, ...  

[1000 rows x 7 columns]

## <FONT SIZE=5 COLOR="magenta"> 4.2 Usando Repositorio Externo url </FONT></center>

Para importar datos desde un repositorio externo, debemos conocer la *url*.

https://archive.ics.uci.edu/datasets?skip=10&take=10&sort=desc&orderBy=NumHits&search=

https://github.com/stedy/Machine-Learning-with-R-datasets

In [ ]:
url1 = "https://raw.githubusercontent.com/stedy/Machine-Learning-with-R-datasets/master/credit.csv"
credito = pd.read_csv(url1)
credito

checking_balance  months_loan_duration credit_history     purpose  amount  \
0             < 0 DM                     6       critical    radio/tv    1169   
1         1 - 200 DM                    48         repaid    radio/tv    5951   
2            unknown                    12       critical   education    2096   
3             < 0 DM                    42         repaid   furniture    7882   
4             < 0 DM                    24        delayed   car (new)    4870   
..               ...                   ...            ...         ...     ...   
995          unknown                    12         repaid   furniture    1736   
996           < 0 DM                    30         repaid  car (used)    3857   
997          unknown                    12         repaid    radio/tv     804   
998           < 0 DM                    45         repaid    radio/tv    1845   
999       1 - 200 DM                    45       critical  car (used)    4576   

    savings_balance employment_length  installment_rate personal_status  \
0           unknown           > 7 yrs                 4     single male   
1          < 100 DM         1 - 4 yrs                 2          female   
2          < 100 DM         4 - 7 yrs                 2     single male   
3          < 100 DM         4 - 7 yrs                 2     single male   
4          < 100 DM         1 - 4 yrs                 3     single male   
..              ...               ...               ...             ...   
995        < 100 DM         4 - 7 yrs                 3          female   
996        < 100 DM         1 - 4 yrs                 4   divorced male   
997        < 100 DM           > 7 yrs                 4     single male   
998        < 100 DM         1 - 4 yrs                 4     single male   
999    101 - 500 DM        unemployed                 3     single male   

    other_debtors  ...                  property age  installment_plan  \
0            none  ...               real estate  67              none   
1            none  ...               real estate  22              none   
2            none  ...               real estate  49              none   
3       guarantor  ...  building society savings  45              none   
4            none  ...              unknown/none  53              none   
..            ...  ...                       ...  ..               ...   
995          none  ...               real estate  31              none   
996          none  ...  building society savings  40              none   
997          none  ...                     other  38              none   
998          none  ...              unknown/none  23              none   
999          none  ...                     other  27              none   

      housing existing_credits  default  dependents  telephone foreign_worker  \
0         own                2        1           1        yes            yes   
1         own                1        2           1       none            yes   
2         own                1        1           2       none            yes   
3    for free                1        1           2       none            yes   
4    for free                2        2           2       none            yes   
..        ...              ...      ...         ...        ...            ...   
995       own                1        1           1       none            yes   
996       own                1        1           1        yes            yes   
997       own                1        1           1       none            yes   
998  for free                1        2           1        yes            yes   
999       own                1        1           1       none            yes   

                         job  
0           skilled employee  
1           skilled employee  
2         unskilled resident  
3           skilled employee  
4           skilled employee  
..                       ...  
995       unskilled resident  
996  mangement self-

In [ ]:
url2 = "https://raw.githubusercontent.com/Fabian830348/Bases_Datos/master/Homicidios.csv"
homicidios = pd.read_csv(url2)
homicidios

Fecha Departamento            Municipio      Día      Hora  \
0        1/1/2017     AMAZONAS         LETICIA (CT)  Domingo  15:05:00   
1        1/1/2017    ANTIOQUIA              BETANIA  Domingo  21:20:00   
2        1/1/2017    ANTIOQUIA           COPACABANA  Domingo   4:48:00   
3        1/1/2017    ANTIOQUIA              EBÉJICO  Domingo   5:00:00   
4        1/1/2017    ANTIOQUIA             EL BAGRE  Domingo   6:30:00   
...           ...          ...                  ...      ...       ...   
11960  12/31/2017        VALLE            CALI (CT)  Domingo   0:31:00   
11961  12/31/2017        VALLE              CARTAGO  Domingo  17:12:00   
11962  12/31/2017        VALLE          LA VICTORIA  Domingo  21:20:00   
11963  12/31/2017      VICHADA  PUERTO CARREÑO (CT)  Domingo   2:00:00   
11964  12/31/2017      VICHADA  PUERTO CARREÑO (CT)  Domingo   2:00:00   

                  Barrio    Zona         Clase de sitio  \
0           KILOMETRO 11   RURAL              CARRETERA   
1              LA CUMBRE  URBANA          VIAS PUBLICAS   
2                MACHADO  URBANA  DENTRO DE LA VIVIENDA   
3                  SAGUA   RURAL     FINCAS Y SIMILARES   
4             LAS BRISAS  URBANA          VIAS PUBLICAS   
...                  ...     ...                    ...   
11960  EL POBLADO II E13  URBANA          VIAS PUBLICAS   
11961    URB. BELLAVISTA  URBANA          VIAS PUBLICAS   
11962             CENTRO  URBANA          VIAS PUBLICAS   
11963           ACEITICO   RURAL                   ISLA   
11964           ACEITICO   RURAL                   ISLA   

                     Arma empleada         Móvil Agresor  \
0      ARMA BLANCA / CORTOPUNZANTE  PASAJERO MOTOCICLETA   
1                    ARMA DE FUEGO                 A PIE   
2      ARMA BLANCA / CORTOPUNZANTE                 A PIE   
3                    ARMA DE FUEGO                 A PIE   
4      ARMA BLANCA / CORTOPUNZANTE                 A PIE   
...                            ...                   ...   
11960                ARMA DE FUEGO                 A PIE   
11961  ARMA BLANCA / CORTOPUNZANTE                 A PIE   
11962  ARMA BLANCA / CORTOPUNZANTE                 A PIE   
11963  ARMA BLANCA / CORTOPUNZANTE                 A PIE   
11964  ARMA BLANCA / CORTOPUNZANTE                 A PIE   

              Móvil Victima  Edad       Sexo Estado civil País de nacimiento  \
0      PASAJERO MOTOCICLETA    20  MASCULINO      SOLTERO           COLOMBIA   
1                     A PIE    21  MASCULINO      SOLTERO           COLOMBIA   
2                     A PIE    22  MASCULINO  UNION LIBRE           COLOMBIA   
3                     A PIE    46  MASCULINO      SOLTERO           COLOMBIA   
4                     A PIE    37  MASCULINO  UNION LIBRE           COLOMBIA   
...                     ...   ...        ...          ...                ...   
11960                 A PIE    19  MASCULINO      SOLTERO           COLOMBIA   
11961                 A PIE    30  MASCULINO      SOLTERO           COLOMBIA   
11962                 A PIE    40   FEMENINO      SOLTERO           COLOMBIA   
11963                 A PIE    45  MASCULINO      SOLTERO           COLOMBIA   
11964                 A PIE    54   FEMENINO      SOLTERO           COLOMBIA   

         Clase de empleado Profesión Escolaridad Código DANE  Cantidad  
0      EMPLEADO PARTICULAR         -  SECUNDARIA  91,001,000         1  
1            INDEPENDIENTE         -    PRIMARIA   5,091,000         1  
2            INDEPENDIENTE         -  SECUNDARIA   5,212,000         1  
3            INDEPENDIENTE         -  SECUNDARIA   5,240,000         1  
4            INDEPENDIENTE         -    PRIMARIA   5,250,000         1  
...                    ...       ...         ...         ...       ...  
11960  EMPLEADO PARTICULAR         -  SECUNDARIA  76,001,000         1  
11961          DESEMPLEADO         -    PRIMARIA  76,147,000         1  
11962          AMA DE CASA         -  SECUNDARIA  76,403,000         1  
11

In [ ]:
diabetes = pd.read_csv("https://raw.githubusercontent.com/Fabian830348/Bases_Datos/refs/heads/master/diabetes.csv ")
diabetes

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [ ]:
comprar = pd.read_csv("https://raw.githubusercontent.com/jbagnato/machine-learning/refs/heads/master/comprar_alquilar.csv")
comprar

ingresos  gastos_comunes  pago_coche  gastos_otros  ahorros  vivienda  \
0        6000            1000           0           600    50000    400000   
1        6745             944         123           429    43240    636897   
2        6455            1033          98           795    57463    321779   
3        7098            1278          15           254    54506    660933   
4        6167             863         223           520    41512    348932   
..        ...             ...         ...           ...      ...       ...   
197      3831             690         352           488    10723    363120   
198      3961            1030         270           475    21880    280421   
199      3184             955         276           684    35565    388025   
200      3334             867         369           652    19985    376892   
201      3988            1157         105           382    11980    257580   

     estado_civil  hijos  trabajo  comprar  
0               0      2        2        1  
1               1      3        6        0  
2               2      1        8        1  
3               0      0        3        0  
4               0      0        3        1  
..            ...    ...      ...      ...  
197             0      0        2        0  
198             2      3        8        0  
199             1      3        8        0  
200             1      2        5        0  
201             0      0        4        0  

[202 rows x 10 columns]

In [ ]:
aptos

In [ ]:
datos = pd.read_json("https://raw.githubusercontent.com/joanby/python-ml-course/refs/heads/master/datasets/titanic/titanic_custom.json")
datos

Para lo que sigue vamos a usar el conjunto **apartamentos**

1. Usamos la instrucción **head()** para explorar la cabeza de los datos

- Podemos indicar cuántos datos queremos visualizar simplemente con $head(n)$, sonde $n$ indica el número de registros que queremos ver.

In [ ]:
# para visualizar los primeros datos.
aptos.head(4)

Unnamed: 0  precio    mt2 ubicacion  estrato  alcobas  banos balcon  \
0           1    79.0  43.16     norte        3        3      1     si   
1           2    93.0  56.92     norte        2        2      1     si   
2           3   100.0  66.40     norte        3        2      2     no   
3           4   123.0  61.85     norte        2        3      2     si   

  parqueadero  administracion     avaluo terminado  
0          si           0.050  14.923002        no  
1          si           0.069  27.000000        si  
2          no           0.000  15.738427        no  
3          si           0.130  27.000000        no

2. Usamos la instrucción **tail()** para explorar los últimos datos

In [ ]:
# para visualizar los últimos datos. (n=6)
aptos.tail(6)

Unnamed: 0  precio    mt2 ubicacion  estrato  alcobas  banos balcon  \
688         689   470.0  248.0  laureles        5        3      2     no   
689         690   498.0  180.0  laureles        5        4      3     si   
690         691   508.0  209.0  laureles        6        4      3     si   
691         692   550.0  149.0  laureles        6        3      4     no   
692         693   630.0  179.0  laureles        5        3      3     si   
693         694  1350.0  494.0  laureles        5        3      4     si   

    parqueadero  administracion   avaluo terminado  
688          si          0.4809  345.376        si  
689          si          0.5700  233.784        si  
690          si          0.4500  258.850        si  
691          si          0.7470  326.157        si  
692          si          0.3400  324.114        si  
693          si          1.6000  320.000        si

3. Con la instrucción **shape** miramos el tamaño de nuestros datos

In [ ]:
# tamaño de los datos
aptos.shape

(694, 12)

el conjunto tiene 694 filas y 12 columnas.

4. Para explorar nombre de las variables o columnas **columns**

In [ ]:
# columnas del conjunto de datos
aptos.columns

Index(['Unnamed: 0', 'precio', 'mt2', 'ubicacion', 'estrato', 'alcobas',
       'banos', 'balcon', 'parqueadero', 'administracion', 'avaluo',
       'terminado'],
      dtype='object')

5. **info()** es un resumen de las variables y su clasificación. Además, podemos identificar datos faltantes (NA) en cada una de ellas.

In [ ]:
# información de los datos
aptos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 694 entries, 0 to 693
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      694 non-null    int64  
 1   precio          694 non-null    float64
 2   mt2             694 non-null    float64
 3   ubicacion       694 non-null    object 
 4   estrato         694 non-null    int64  
 5   alcobas         694 non-null    int64  
 6   banos           694 non-null    int64  
 7   balcon          694 non-null    object 
 8   parqueadero     694 non-null    object 
 9   administracion  694 non-null    float64
 10  avaluo          694 non-null    float64
 11  terminado       694 non-null    object 
dtypes: float64(4), int64(4), object(4)
memory usage: 65.2+ KB


Es importante verificar que las variables del conjunto de datos hayan sido cargadas en el tipo adecuado.

6. Para revisar las estadísticas generales de las variables cuantitativas usamos **describe()**

In [ ]:
# estadísticas descriptivas
aptos.describe()

Unnamed: 0       precio         mt2     estrato     alcobas  \
count  694.000000   694.000000  694.000000  694.000000  694.000000   
mean   347.500000   317.762739  120.943818    4.651297    2.910663   
std    200.484829   247.614914   74.695493    1.208205    0.763962   
min      1.000000    25.000000   26.000000    2.000000    1.000000   
25%    174.250000   160.000000   71.000000    4.000000    3.000000   
50%    347.500000   245.000000   97.500000    5.000000    3.000000   
75%    520.750000   380.000000  141.750000    6.000000    3.000000   
max    694.000000  1700.000000  500.000000    6.000000   14.000000   

            banos  administracion       avaluo  
count  694.000000      694.000000   694.000000  
mean     2.289625        0.263556   182.636593  
std      0.795718        0.251586   169.082384  
min      1.000000        0.000000     0.149000  
25%      2.000000        0.090000    73.000000  
50%      2.000000        0.201000   131.431500  
75%      3.000000        0.375000   234.051000  
max      6.000000        2.280000  1540.620000

In [ ]:
# si queremos las estadísticas (frecuencias) de las variables cualitativas
aptos.describe(include='object')

ubicacion balcon parqueadero terminado
count        694    694         694       694
unique         7      2           2         2
top      poblado     si          si        si
freq         268    509         568       592

7. Para acceder a una o varias variable del data.frame podemos usar la instrucción

*datos.["nombre de la variable"]*

veamos algunos ejemplos

In [ ]:
# podemos acceder a una o más variables del data.frame data[]
aptos["precio"]

In [ ]:
aptos.columns

In [ ]:
aptos[["ubicacion","estrato","banos"]]

ubicacion  estrato  banos
0       norte        3      1
1       norte        2      1
2       norte        3      2
3       norte        2      2
4       norte        4      2
..        ...      ...    ...
689  laureles        5      3
690  laureles        6      3
691  laureles        6      4
692  laureles        5      3
693  laureles        5      4

[694 rows x 3 columns]

In [ ]:
homicidios.columns

Index(['Fecha', 'Departamento', 'Municipio', 'Día', 'Hora', 'Barrio', 'Zona',
       'Clase de sitio', 'Arma empleada', 'Móvil Agresor', 'Móvil Victima',
       'Edad', 'Sexo', 'Estado civil', 'País de nacimiento',
       'Clase de empleado', 'Profesión', 'Escolaridad', 'Código DANE',
       'Cantidad'],
      dtype='object')

In [ ]:
homicidios["Arma empleada"].unique()

array(['ARMA BLANCA / CORTOPUNZANTE', 'ARMA DE FUEGO', 'CONTUNDENTES',
       'CUERDA/SOGA/CADENA', 'LIQUIDOS', 'COMBUSTIBLE', 'CINTAS/CINTURON',
       'SIN EMPLEO DE ARMAS', 'MINA ANTIPERSONA', 'GRANADA DE MANO',
       'ARTEFACTO EXPLOSIVO/CARGA DINAMITA', 'VENENO',
       'ARTEFACTO INCENDIARIO', 'BOLSA PLASTICA', 'SUSTANCIAS TOXICAS',
       'ACIDO', 'ALMOHADA', 'QUIMICOS', 'ESCOPOLAMINA',
       'PRENDAS DE VESTIR'], dtype=object)

In [ ]:
aptos.ubicacion.unique()

array(['norte', 'occidente', 'poblado', 'aburra sur', 'belen guayabal',
       'centro', 'laureles'], dtype=object)

In [ ]:
# podemos revisar unas estadísticas básicas de una variable numérica con .describe()
aptos["precio"].describe()

count     694.000000
mean      317.762739
std       247.614914
min        25.000000
25%       160.000000
50%       245.000000
75%       380.000000
max      1700.000000
Name: precio, dtype: float64

In [ ]:
aptos[["precio", "mt2"]]

precio     mt2
0      79.0   43.16
1      93.0   56.92
2     100.0   66.40
3     123.0   61.85
4     135.0   89.80
..      ...     ...
689   498.0  180.00
690   508.0  209.00
691   550.0  149.00
692   630.0  179.00
693  1350.0  494.00

[694 rows x 2 columns]

También se puede acceder al nombre de una variable asi:
$$datos.variable$$
Siempre y cuando el nombre de la variable no tenga espacios o caracteres especiales.

In [ ]:
aptos.precio

0        79.0
1        93.0
2       100.0
3       123.0
4       135.0
        ...  
689     498.0
690     508.0
691     550.0
692     630.0
693    1350.0
Name: precio, Length: 694, dtype: float64

# <FONT SIZE=5 COLOR="green"> 5. Tabla de frecuencias </FONT>

Para hacer el conteo de las categorías o clases de una variables podemos usar la instrucción ***.value_counts***

Ejemplo: Para hacer una tabla de frecuencias de la variable *ubicacion*

In [ ]:
aptos["ubicacion"].value_counts()

ubicacion
poblado           268
aburra sur        169
laureles           73
occidente          69
belen guayabal     67
centro             38
norte              10
Name: count, dtype: int64

In [ ]:
# conteo de la variable terminado-aptos
aptos["terminado"].value_counts()

terminado
si    592
no    102
Name: count, dtype: int64

In [ ]:
homicidios.columns

Index(['Fecha', 'Departamento', 'Municipio', 'Día', 'Hora', 'Barrio', 'Zona',
       'Clase de sitio', 'Arma empleada', 'Móvil Agresor', 'Móvil Victima',
       'Edad', 'Sexo', 'Estado civil', 'País de nacimiento',
       'Clase de empleado', 'Profesión', 'Escolaridad', 'Código DANE',
       'Cantidad'],
      dtype='object')

In [ ]:
# ¿cuántas personas fueron asesinas empleando veneno? : homicidios
homicidios["Arma empleada"].unique()

array(['ARMA BLANCA / CORTOPUNZANTE', 'ARMA DE FUEGO', 'CONTUNDENTES',
       'CUERDA/SOGA/CADENA', 'LIQUIDOS', 'COMBUSTIBLE', 'CINTAS/CINTURON',
       'SIN EMPLEO DE ARMAS', 'MINA ANTIPERSONA', 'GRANADA DE MANO',
       'ARTEFACTO EXPLOSIVO/CARGA DINAMITA', 'VENENO',
       'ARTEFACTO INCENDIARIO', 'BOLSA PLASTICA', 'SUSTANCIAS TOXICAS',
       'ACIDO', 'ALMOHADA', 'QUIMICOS', 'ESCOPOLAMINA',
       'PRENDAS DE VESTIR'], dtype=object)

In [ ]:
homicidios["Arma empleada"].value_counts()

Si queremos una tabla cruzada entre dos variables categóricas usamos la instrucción:

$$pd.crosstab$$

In [ ]:
pd.crosstab(aptos.ubicacion , aptos.estrato)

estrato         2   3   4   5    6
ubicacion                         
aburra sur      3  73  57  35    1
belen guayabal  1  30  21  15    0
centro          0  16  21   1    0
laureles        0   0  14  56    3
norte           2   5   3   0    0
occidente       2  37  15  15    0
poblado         0   0   7  23  238

In [ ]:
homicidios.columns

Index(['Fecha', 'Departamento', 'Municipio', 'Día', 'Hora', 'Barrio', 'Zona',
       'Clase de sitio', 'Arma empleada', 'Móvil Agresor', 'Móvil Victima',
       'Edad', 'Sexo', 'Estado civil', 'País de nacimiento',
       'Clase de empleado', 'Profesión', 'Escolaridad', 'Código DANE',
       'Cantidad'],
      dtype='object')

In [ ]:
# tabla de frecuencia (contingencia) ubicacion y parqueadero
pd.crosstab(aptos["ubicacion"],                 # variable 1
            aptos["parqueadero"],               # variable 2
            margins= True,                      # muestra el total
            margins_name= "SubTotal")           # nombre del total

parqueadero      no   si  SubTotal
ubicacion                         
aburra sur       39  130       169
belen guayabal   31   36        67
centro           28   10        38
laureles          3   70        73
norte             3    7        10
occidente        20   49        69
poblado           2  266       268
SubTotal        126  568       694

In [ ]:
pd.crosstab(homicidios["Sexo"],                 # variable 1
            homicidios["Estado civil"],               # variable 2
            margins= True,                      # muestra el total
            margins_name= "SubTotal")           # nombre del total

Estado civil   -  CASADO  DIVORCIADO  SEPARADO  SOLTERO  UNION LIBRE  VIUDO  \
Sexo                                                                          
FEMENINO       3      71          11        18      552          385     21   
MASCULINO     45     558          28        60     6636         3547     30   
SubTotal      48     629          39        78     7188         3932     51   

Estado civil  SubTotal  
Sexo                    
FEMENINO          1061  
MASCULINO        10904  
SubTotal         11965

In [ ]:
# tabla de contingencia de tres variables cruzadas
pd.crosstab(aptos["ubicacion"], [aptos["balcon"], aptos["estrato"]])

balcon         no                 si                 
estrato         2   3   4   5   6  2   3   4   5    6
ubicacion                                            
aburra sur      2  26   6   4   0  1  47  51  31    1
belen guayabal  0   6   5   1   0  1  24  16  14    0
centro          0   8  10   0   0  0   8  11   1    0
laureles        0   0   5  18   2  0   0   9  38    1
norte           0   2   0   0   0  2   3   3   0    0
occidente       1  17   7   3   0  1  20   8  12    0
poblado         0   0   2   9  51  0   0   5  14  187

# <FONT SIZE=5 COLOR="green"> 6. Manipulación de DataFrame </FONT>

En esta sección revisaremos algunas operaciones y acciones que son importantes de conocer en un Data.Frame:

- Ordenar por una variable

- Renombrar una variable

- Eliminar una variable

- Agregar una variable

- Agrupar por categorias de una variable y aplicar funciones.

Para repasar

https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf

## <FONT SIZE=4 COLOR="magenta"> 6.1 Ordenar según una variable </FONT>

Para ordenar por una variable usamos la instrucción

$$datos.sort\_values(variable, ascending = False) $$

se usa:

- *ascending=True* :  ordena de menor a mayor. (default)
- *ascending=False* : ordena de mayor a menor.

In [ ]:
# ordenar de mayor a menor la edad en homicidios
homicidios.sort_values("Edad", ascending = False)

In [ ]:
# default ordena de menor a mayor
aptos.sort_values("ubicacion", ascending = False)

Unnamed: 0  precio    mt2   ubicacion  estrato  alcobas  banos balcon  \
209         210   400.0  137.0     poblado        6        3      3     si   
222         223   430.0  108.0     poblado        6        2      3     si   
230         231   440.0  106.0     poblado        6        3      2     si   
229         230   440.0  176.0     poblado        6        3      2     si   
228         229   440.0  202.0     poblado        6        3      3     si   
..          ...     ...    ...         ...      ...      ...    ...    ...   
487         488   265.0   81.0  aburra sur        5        3      2     si   
488         489   265.0   74.0  aburra sur        6        2      2     si   
489         490   265.0   72.0  aburra sur        5        2      2     si   
490         491   265.0   72.0  aburra sur        5        2      2     si   
347         348    45.0   53.0  aburra sur        3        3      1     si   

    parqueadero  administracion      avaluo terminado  
209          si        0.358200  202.202000        no  
222          si        0.330000  355.089000        no  
230          si        0.330000  284.000000        si  
229          si        0.450000  336.000000        si  
228          si        0.421402  301.590000        si  
..          ...             ...         ...       ...  
487          si        0.197000   70.050464        si  
488          si        0.266200   97.775997        si  
489          si        0.183000   70.050464        si  
490          si        0.183000   79.709667        si  
347          no        0.000000   23.899000        si  

[694 rows x 12 columns]

In [ ]:
# ordenamos por la variable Edad de mayor a menor
aptos.sort_values("precio" , ascending =  False)

In [ ]:
# ordenamos por la variable Edad de menor a mayor
aptos.sort_values("precio" , ascending =  True)

Unnamed: 0  precio     mt2   ubicacion  estrato  alcobas  banos balcon  \
644         645    25.0  118.00    laureles        5        3      2     si   
347         348    45.0   53.00  aburra sur        3        3      1     si   
10           11    47.0   43.00   occidente        2        2      1     no   
11           12    53.0   37.56   occidente        3        3      1     no   
348         349    63.0   43.00  aburra sur        3        3      1     no   
..          ...     ...     ...         ...      ...      ...    ...    ...   
342         343  1430.0  291.00     poblado        6        3      3     si   
344         345  1450.0  238.19     poblado        6        4      4     si   
343         344  1450.0  318.00     poblado        6        3      3     si   
345         346  1490.0  445.00     poblado        6        5      4     si   
346         347  1700.0  399.00     poblado        6        3      3     no   

    parqueadero  administracion   avaluo terminado  
644          no          0.0000   95.885        si  
347          no          0.0000   23.899        si  
10           no          0.0000   16.486        no  
11           si          0.0000   16.486        si  
348          no          0.0500   41.468        si  
..          ...             ...      ...       ...  
342          si          0.7490  701.731        si  
344          si          0.8286  675.072        si  
343          si          0.7490  611.027        si  
345          si          0.7490  701.731        si  
346          si          0.9340  701.731        si  

[694 rows x 12 columns]

## <FONT SIZE=4 COLOR="magenta"> 6.2 Renombrar una variable </FONT>

Podemos renombrar una columna de la siguiente forma usando la instrucción ***rename***

$$ datos.rename(columns=\{key : value, key : value\}, inplace=True) $$

 - Para cambiar el nombre de una variable debemos usar un diccionario donde las claves son las variables actuales y los valores son los nombres nuevos de las variables.
 - Recordemos que la instrucción **inplace** se usa para determinar si cambiamos el DataFrame original.

In [ ]:
# cambiaremos el nombre de administracion por admon
aptos.rename(columns={" " : " "}, inplace=False)

## <FONT SIZE=4 COLOR="magenta"> 6.3 Eliminar filas y columnas</FONT>

Para eliminar una columna o una fila usamos la instrucción .drop

$$datos.drop(``nombre", axis = 1, inplace = T \, or \, F)$$

- *axis = 1* : es por columna
- *axis = 0* : es por  fila
- *inplace = T* : sobreescribe en el *df* original.
- *inplace = F* : no sobreescribe. Mantiene el original

In [ ]:
# vamos a crear una copia del dataframe
aptos_copy = aptos.copy()

In [ ]:
# axis = 1
aptos_copy.drop("Unnamed: 0", axis = 1 , inplace = True)

In [ ]:
aptos_copy.head()

<FONT SIZE=4 COLOR="magenta"> 6.4 Agregar una variable </FONT>

Para agregar una variable usamos la instrucción

$$datos[``\text{nueva variable}"] = lista $$

In [ ]:
aptos["precio/mt2"]= aptos["precio"]/aptos["mt2"]


In [ ]:
aptos

In [ ]:
aptos.precio.mean()

317.7627391930836

In [ ]:
aptos.groupby("ubicacion")["precio"].describe()

count        mean         std    min     25%    50%     75%  \
ubicacion                                                                     
aburra sur      169.0  193.672781   64.570992   45.0  140.00  185.0  250.00   
belen guayabal   67.0  168.865672   50.277983   75.0  140.00  165.0  210.00   
centro           38.0  130.815789   50.874438   67.0   87.00  128.0  149.25   
laureles         73.0  318.520548  159.220594   25.0  235.00  290.0  370.00   
norte            10.0  131.000000   31.770007   79.0  105.75  137.5  156.25   
occidente        69.0  164.304348   59.968694   47.0  120.00  160.0  210.00   
poblado         268.0  506.017317  290.683116  140.0  320.00  415.0  605.00   

                   max  
ubicacion               
aburra sur       330.0  
belen guayabal   265.0  
centro           260.0  
laureles        1350.0  
norte            175.0  
occidente        320.0  
poblado         1700.0

## <FONT SIZE=4 COLOR="magenta"> 6.5 Agrupaciones: ***groupby()*** </FONT>



In [ ]:
aptos.groupby("ubicacion")

Observe que no se genera nada. Esto es porque internamente lo que hace pandas es agrupar por las clases o categorias de la variable.

**¿Qué podemos hacer con esto?**

Podemos hacer operaciones con las agrupaciones, así:

$$datos.groupby(``Variable")[``\text{segunda variable}"].funcion()$$

Por ejemplo, supongamos que queremos agrupar por escolaridad y calcular el promedio de las edades de cada uno de los grupos.

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/groupby2.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# calcula la media del precio en cada ubicación
aptos.groupby("ubicacion")["precio"].mean()

In [ ]:
# calcular las estadísticas del precio en cada ubicación
aptos.groupby("ubicacion")["precio"].describe()

In [ ]:
# si agrupamos por ubicacion, y parqueadero quiero saber el valor máximo del precio
aptos.groupby(["parqueadero" , "terminado"]).precio.max()

In [ ]:
aptos.groupby(["ubicacion", "parqueadero"]).precio.min()

# <FONT SIZE=5 COLOR="green"> 7. Algunas funciones importantes </FONT>

1. Media o promedio: **.mean()**

- Es el "equilibrio" de los datos.
- Se usa para variables cuantitativas.
- Siempre está entre el mínimo y máximo de los datos.
- Es sensible a datos atípicos. Por está razón se debe tener cuidado con su interpretación.

In [ ]:
aptos.precio.mean()

2. Mediana: **.median()**

- Es el valor que divide los datos en dos partes iguales (50%-50%)
- Aplica para variables cuantitativas
- Puede o no estar en el conjunto de datos. Depende si se tiene un número par o impar de elementos.
- No se ve afectada por la existencia de datos atípicos. Por está razón en ocaciones se usa para imputación.

In [ ]:
aptos.precio.median()

3. Moda: **.mode()**

- Es el valor que más se repite. Es decir el que tiene mayor frecuencia
- Aplica para variables cualitativas y cuantitativas
- UN conjunto puede no tener o tener múltiples modas.

In [ ]:
aptos.ubicacion.mode()

4. Máximo y mínimo del conjunto: **.max** y **.min**

In [ ]:
aptos.precio.max()

In [ ]:
aptos.precio.min()

5. Cuartiles: $Q_1$, $Q_2$ y $Q_3$

- Se calculan para variables cuantitativas
- $Q_1$: el valor que acumula el 25% de los datos
- $Q_2$: es el valor que acumula el 50% de los datos. En ese orden de ideas es la mediana.
- $Q_3$ es el valor que acumula el 75% de los datos.
- Para calcula en Python estos valores y en general los cuantiles usamos la librería *numpy* y la instrucción:

$$ perc\_{n} = np.percentile(variable, n) $$

Veamos algunos ejemplos con la variable *precio* de *aptos*.

In [ ]:
import numpy as np
print(np.percentile(aptos.precio,25))
print(np.percentile(aptos.precio,50))
print(np.percentile(aptos.precio,75))
print(np.percentile(aptos.precio,80))

# <FONT SIZE=5 COLOR="green"> 8. Algunas gráficas importantes </FONT>

En esta sección revisaremos algunas gráficas clásicas.

Para esto podemos usar diferentes librerías: *mathplotlib* y *plotly*. También se pueden hacer gráficos elaborados en *seaborn* Más documentación la pueden encontrar en los siguientes links.

- https://matplotlib.org/

- https://seaborn.pydata.org/

- https://plotly.com/python/plotly-express/

**Nota:** nos centraremos más en la parte gráfica y el código que en el análisis de las gráficas.

<center><FONT SIZE=5 COLOR="Magenta"> Mathplotlib - Seaborn - Plotly.Express </FONT></center>

Existen múltiples librerías para hacer gráficas en Python. En esta parte mostraremos algunos ejemplos, sin embargo, invitamos al lector a explorar con más detalle estas librerías.

- **Mathplotlib**: es una biblioteca de Python utilizada para trazar gráficos con la ayuda de otras bibliotecas como Numpy y Pandas. Es una poderosa herramienta para visualizar datos en Python. Se utiliza para crear inferencias estáticas y trazar gráficos 2D de matrices. Fue presentado por primera vez por John D. Hunter en 2002.


- **Seaborn**: también es una biblioteca de Python que se utiliza para trazar gráficos con la ayuda de Matplotlib, Pandas y Numpy. Está construido sobre el techo de Matplotlib y se considera como un superconjunto de la biblioteca Matplotlib. Ayuda a visualizar datos univariados y bivariados.

- **Plotly.express** es otra librería para visualización con gráficos interactivos que muestran algunos detalles de las gráficas.

Vamos a cargar las librerías

In [ ]:
import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sb

## <FONT SIZE=4 COLOR="green"> 8.1. TABLA DE FRECUENCIA - BARRAS Y TORTAS </FONT>

Para hacer el gráfico de barras o de tortas haremos primero una tabla de frecuencias.

- Recordemos de los gráficos de barras o de tortas se usan en general para variables categóricas o para variables cuantitativas con pocos valores, como es el caso de *número de hijos*.

- En el eje $x$ van las categorías

- En el eje $y$ la cantidad o frecuencia de cada clase o categoria

- En los diagramas de barras las barras van separadas y de colocan de diferente color.

Vamos a hacer la tabla de frecuencias absolutas y los diagramas de barras y tortas de la variable *ubicacion* del conjunto de datos aparatamentos.

In [ ]:
freq = pd.DataFrame(aptos.ubicacion.value_counts())
freq

Vamos a cambiar el nombre a la columna y dejarla frecuencia absoluta (freq_abs)

In [ ]:
freq.rename(columns={'count':'freq_abs'}, inplace = True)
freq

In [ ]:
# gráfico de barras
px.bar(x = freq.index,
       y= freq.freq_abs,
       color = freq.index)

In [ ]:
# A partir de la tabla de frecuencias de la variable ubicación
px.pie(freq,                               # tabla de frecuencias
       values= freq.freq_abs,              # valores de las frecuencias absolutas
       names= freq.index)                  # nombres de las categorias

## <FONT SIZE=4 COLOR="green"> 8.2 HISTOGRAMAS ABSOLUTO Y ACUMULADO </FONT>

- El histograma se usa para variables de tipo cuantitativo continuo o discreto con muchas mediciones diferentes (como la edad en ciertos casos).

- En el eje $x$ se identifican las clases (intervalos) de la variable.

- En el eje $y$ se identifica la frecuencia de cada una de las clases.

- Las barras en el histograma quedan contiguas y además se utiliza un solo color en estas.

In [ ]:
# Histograma más básico
px.histogram(aptos.precio)

In [ ]:
fig = px.histogram(aptos.precio,                            # variable
                   nbins = 30,                              # número de intervalos
                   x = "precio",                            # nombre de la variable x
                   title = "Histograma con edición"         # título del gráfico
                                      )

fig.update_traces(marker_color = 'cyan',                    # color de relleno
                  marker_line_color = 'red',                # color del borde
                  marker_line_width = 1,                    # grueso de la línea
                  opacity = 1)                              # transparencia del color

fig.show()

## <FONT SIZE=4 COLOR="green"> 8.3. BOXPLOT (GRÁFICO DE CAJAS Y BIGOTES) </FONT>

- El gráfico de caja, boxplot del inglés, fue ideado por Jhon Tukey en 1977 y es una representación estadística que resalta principalmente aspectos de la distribución de los datos u observaciones por medio de las medidas de posición.

- Los diagramas de caja brindan información valiosa que no se limpia en un histograma o en un polígono de frecuencias.

A continuación a describir cada uno de los elementos que tiene un boxplot

**La caja rectangular** : el gráfico contiene una caja rectangular. Es importante indicar que esta caja no tiene coordenadas cartesianas en el eje $X$  y las alturas en el eje $y$ no representa una escala de frecuencias. Las lineas horizontales en el diagrama de cajas describen el primer y tercer cuartil. $Q_1$ y $Q_3$ respectivamente.

**Mediana** : es una línea que se muestra dentro de la caja rectangular y hace referencia al valor que divide los datos en dos partes iguales. Recuerde que la mediana es igual a $Q_2$.

**Bigotes** : son lineas que salen a los costados de la caja y se usan como referencia para ubicar las observaciones que estan por fuera del $50 \%$ central de distribucion. Los bigotes nos dan información sobre la existencia de valores atípicos (*outliers*).

In [ ]:
# boxplot clásico
px.box(aptos.mt2)

In [ ]:
px.box(aptos,                                           # nombre de los datos
       y = "precio",                                    # variable cuantitativa
       x = "ubicacion",                                 # variable cualitativa o categórica
       title = "Boxplot variable precio vs ubicación",  # título del gráfico
       color = "ubicacion"                              # color por la variable categórica
             )

## <FONT SIZE=4 COLOR="green"> 8.4 DIAGRAMA DE DISPERSIÓN </FONT>

- Los diagramas de dispersión permiten representar graficamente dos variables cuantitativas.

In [ ]:
px.scatter(aptos,
           x= "precio",
           y= "mt2")

In [ ]:
px.scatter(aptos,                                     # datos
           x= "precio",                               # variable 1
           y= "mt2",                                  # variable 2
           color = "ubicacion")

In [ ]:
px.scatter(aptos,
           x= "precio",
           y= "mt2",
           color = "ubicacion",
           size = "avaluo")